### ISSUE TITLE
Error: Not found. Authentication passthru. with any Omniauth strategy

### ISSUE THREAD

**Issue by lionelrudaz** [heartcombo/devise#5236]
Hi everybody, I'm posting here, because none of the answers in StackOverflow seem to work. So either I've got a bug with the library, or this issue can lead to an improvement of the documentation.

## Environment

- Ruby 2.7.0
- Rails 6.0.3
- Devise 4.7.1
- Omniauth 1.9.1
- omniauth-facebook 6.0.0

## Current behavior

I've published my repo here: https://github.com/lionelrudaz/dunphy-api/tree/omniauth-facebook

I've followed the guide to enable Facebook login with Omniauth. I've done it three times. I've checked all the possible similar questions on Stackoverflow, including:

- https://stackoverflow.com/questions/41323346/not-found-authentication-passthru-devise-oauth-facebook
- https://stackoverflow.com/questions/17504452/devise-omniauth-facebook-error-not-found-authentication-passthru
- https://stackoverflow.com/questions/13812844/devise-omniauth-facebook-not-found-authentication-passthru
- https://stackoverflow.com/questions/16755889/i-keep-getting-not-found-authentication-passthru-for-facebook-connect-login-w

When I'm entering http://localhost:3000/users/auth/facebook/, I always get the same error.

Looks like I'm not the only one that had struggle with the error. What's strange is that most of the time, the issue either was fixed by redoing everything from scratch, restarting the server (which I did many times), upgrading rails (which I did as well from 6.0.2 to 6.0.3) or by waiting for the app to be enabled...

I've got no logs to provide, except this error.

Will be even more helpful if you provide a sample application or a test case that reproduces the error.

## Expected behavior

I should be redirected to Facebook login. 

Let me know if you need more information.

**Comment by apchester** [comment]
Did you get any further with this? I'm getting the same problem.

**Comment by lionelrudaz** [comment]
No, I didn't. I changed the flow to make it work differently, taking the authorization code from the client, then sending it to my Rails app.

**Comment by arcooverbeek** [comment]
@lionelrudaz and @apchester please make sure you use `method: :post` for you links. Also make sure JS is working. 
`= link_to "Sign in with Facebook", user_facebook_omniauth_authorize_path, method: :post`

**Comment by apchester** [comment]
Thanks! instantly fixes the issue for me, though this doesn't appear to be the default?

**Comment by lionelrudaz** [comment]
@arcooverbeek the thing is that I was in API mode for my Rails app, and my calls were done from a different frontend application. In this case, I don't think it makes sense to post to the URL, don't you think?

**Comment by Pandenok** [comment]
@arcooverbeek using `method: :post` leads to `WARN -- omniauth: Attack prevented by OmniAuth::AuthenticityTokenProtection`. How ti fix `Not found. Authentication passthru.`?

**Comment by carlosantoniodasilva** [comment]
Hey all, please note that using POST as the method is the new default in OmniAuth 2+, and that is to increase security and is definitely recommended.

Devise is upgrading to that default soon: https://github.com/heartcombo/devise/pull/5327, and so should you if possible. Check their release notes for more info: https://github.com/omniauth/omniauth/releases/tag/v2.0.0.

If you're getting to that error it means OmniAuth is not catching the route and doing its thing, and instead you're landing directly in Rails/Devise.

@lionelrudaz are you still encountering the problem you originally reported?

**Comment by Pandenok** [comment]
@carlosantoniodasilva thank you for coming back on this! So, another question, why OmniAuth is not catching the route?
`routes.rb`
```ruby
devise_for :users, controllers: { omniauth_callbacks: 'omniauth_callbacks' }
```

**Comment by carlosantoniodasilva** [comment]
Have you verified that the request is actually being sent as POST with the link using `method: :post`? That requires rails or jQuery UJS, if you're not using that in your app it may still be going through a normal GET request. You can also try changing to a button_to temporarily to test if that is the problem.

With OmniAuth 2.0+, `OmniAuth.config.allowed_request_methods` should default to just include `:post`, while previous versions should have `:get` and `:post`, that may be the reason that it's not working for you.

If that's not the case, it'd be nice if you could provide a sample app that shows the issue. I haven't been able to re-test everything myself yet with that branch. Thanks.

**Comment by jkowens** [comment]
@Pandenok I think you need to set the `OmniAuth.config.request_validation_phase`.

See: https://github.com/omniauth/omniauth/wiki/Upgrading-to-2.0#rails

**Comment by Pandenok** [comment]
@jkowens YEAAAAH! 
1. Install the gem `OmniAuth - Rails CSRF Protection`
2. Add the link user_facebook_omniauth_authorize_path `method: :post`
3. Get the facebook omniauth working without downgrading!
Thank youuuu!
@carlosantoniodasilva sorry for the delay in answering. My bad...

**Comment by carlosantoniodasilva** [comment]
Awesome, I'm happy to know that is working @Pandenok, and thanks for the direction @jkowens.

I'm gonna work on getting a new devise version released this week containing the patches in that branch.

**Comment by Pandenok** [comment]
@carlosantoniodasilva you're doing amazing job!

**Comment by juliaf1** [comment]
@carlosantoniodasilva I have been going through your suggestions on different forums to fix OmniAuth and have landed on an AuthenticityTokenProtection error just like @Pandenok.

I'm using these two gems:
```
gem 'devise', github: 'heartcombo/devise', branch: 'ca-omniauth-2'
gem 'omniauth-facebook'
```

And calling the user_facebook_omniauth_authorize_path with `method: :post`

I tried installing` OmniAuth - Rails CSRF Protection`, but didn't exactly understand where I should set `OmniAuth.config.request_validation_phase`, so that didn't work and I keep landing on this:

```
Started POST "/users/auth/facebook" for ::1 at 2021-01-25 11:56:40 -0300
D, [2021-01-25T11:56:40.532147 #4264] DEBUG -- omniauth: (facebook) Request phase initiated.
W, [2021-01-25T11:56:40.532787 #4264]  WARN -- omniauth: Attack prevented by OmniAuth::AuthenticityTokenProtection
E, [2021-01-25T11:56:40.532940 #4264] ERROR -- omniauth: (facebook) Authentication failure! authenticity_error: OmniAuth::AuthenticityError, Forbidden
Processing by Users::OmniauthCallbacksController#failure as HTML
```

Any thoughts on how to fix this would be super appreciated.

**Comment by jkowens** [comment]
@juliaf1 did you see the Omniauth 2.0 upgrade wiki? https://github.com/omniauth/omniauth/wiki/Upgrading-to-2.0#rails

You can either add `gem "omniauth-rails_csrf_protection"` to your Gemfile or follow the instructions to manually set `OmniAuth.config.request_validation_phase`.

**Comment by carlosantoniodasilva** [comment]
@juliaf1 right, from my understanding the only thing you should need (aside from using that branch and OmniAuth 2) is what @jkowens mentioned above: add the `omniauth-rails_csrf_protection` to your Gemfile, and make sure that's at least `~> 1.0`.

That combined with initiating the OAuth flow using a POST request (using `button_to` or `link_to + method: :post`) should be enough to get it working.

I'll probably include some documentation around this on the Devise side (changelog/readme/wiki), I'm sure more people will hit issues upgrading and will come looking here first.

**Comment by Pandenok** [comment]
@juliaf1 Julia, 
> I tried installing OmniAuth - Rails CSRF Protection, but didn't exactly understand where I should set OmniAuth.config.request_validation_phase, so that didn't work and I keep landing on this
If you installed the gem, you don't need to add anything, it's the gem itself will add request validation on any post request, so make sure as @carlosantoniodasilva stated to user a `link_to + method: :post` or `button_to` which will make by default a post request. 
Please check that you `omniauth` gem is really updated to the latest version by running `bundle omniauth update`

Hope this will help!

**Comment by michaelwebb76** [comment]
If anyone lands here and has just wasted an entire day trying to resolve this issue, re-reading this thread 20 times and screaming "WHY" into their clenched fist, you explicitly need to include `gem "omniauth-rails_csrf_protection", "~> 1.0"` to your *Rails application Gemfile* for the railtie to fire and setup the token verifier correctly - it's not sufficient to have that gem required by another gem if you, like us, have a single "authentication" gem that is used by all of your systems.

**Comment by Flip437** [comment]
Hi,

what worked for me:
- install ge "OmniAuth - Rails CSRF Protection"
- add method: :post to your link_to button.
no more config needed.

peace

**Comment by BryanGoss** [comment]
> If anyone lands here and has just wasted an entire day trying to resolve this issue, re-reading this thread 20 times and screaming "WHY" into their clenched fist, you explicitly need to include `gem "omniauth-rails_csrf_protection", "~> 1.0"` to your _Rails application Gemfile_ for the railtie to fire and setup the token verifier correctly - it's not sufficient to have that gem required by another gem if you, like us, have a single "authentication" gem that is used by all of your systems.

@michaelwebb76 Another route that worked for me is that the "authentication" gem should explicitly `require 'omniauth/rails_csrf_protection'` which also triggers the railtie. The reason adding it to the Gemfile worked is that by default Bundler requires all the gems. It was tricky to get the require statement to work, since the name of the gem and the file structure don't match exactly.

**Comment by seanmsmith23** [comment]
Gems upgraded, method: :post, disable turbo. Same issue. Sigh...

**Comment by viktorianer** [comment]
> Just in case it helps anyone.
> 
> If you got your Turbo enabled and smth like `<%= link_to "Sign in with Facebook", user_facebook_omniauth_authorize_path, method: :post %>` still brings you a CORS-related error, make sure you read [this](https://github.com/hotwired/turbo/issues/45).
> 
> TL;DR: You'd better switch to form instead of a link and set a `turbo` form's param to `false`.

It helps to switch to `button_to`, as long as that is not resolved on Turbo side:
https://github.com/hotwired/turbo/issues/45
https://github.com/hotwired/turbo/issues/176
https://github.com/hotwired/turbo/issues/288

This works for me too:

```ruby
<%= button_to "Sign in with Microsoft", user_microsoft_graph_omniauth_authorize_path,
          method: :post, "data-turbo": false %>
```

**Comment by kg-currenxie** [comment]
This works for me with turbo
```
gem 'omniauth-rails_csrf_protection', '~> 1.0'

and 

<%= link_to user_facebook_omniauth_authorize_path, method: :post, class: 'facebook-button', data: { turbolinks: false }  %>
```
No other changes

**Comment by ragesoss** [comment]
Anyone have advice for working around this problem for the case of "open link in new tab"? My login links work as expected when they are clicked normally, but "method: :post" doesn't prevent this error when a user tries to open the login in a new tab.

**Comment by Duartemartins** [comment]
This is happening to me on Rails 7 using https://github.com/andyw8/omniauth-pocket-oauth2 and none of these solutions seem to work :/.

**Comment by zal-lex** [comment]
Only replacing `<%= link_to "Sign in with #{provider}", user_provider_omniauth_authorize_path, method: :post %>` with ` <%= button_to "Sign in with #{OmniAuth::Utils.camelize(provider)}", user_doorkeeper_omniauth_authorize_path %>` worked for me.

**Comment by carlosantoniodasilva** [comment]
@zal-lex are you using turbo? If so, there's a reason that it works with button_to and not link_to, see https://github.com/rails/rails/pull/43430 and https://github.com/rails/rails/pull/43941

**Comment by bjnord** [comment]
Rails 7 doesn't have UJS by default, so the previous `method: :post` directive on `link_to` doesn't work any more. I found [this](https://stackoverflow.com/a/70504229/291754) on StackOverflow with the answer: Change `method: :post` to `data: { 'turbo-method' => :post }`

**Comment by ketan-survival** [comment]
Thanks @jkowens.

Adding `gem 'omniauth-rails_csrf_protection'` to the Gemfile and Using a `form_tag` with a submit button worked for me, a standalone `button_to` or `link_to` with `method: :post` wouldn't work.

Also `gem 'omniauth-rails_csrf_protection'` pulls version `1.0.1` at the time of this comment so no need to specify version.

```slim
      = form_tag(admin_user_google_oauth2_omniauth_authorize_path) do
        button(type='submit' class='btn btn-primary', style='width: 100%')
          i.fa.fa-google
          |  Sign in with Google
```

**Comment by hs-dliaska** [comment]
if you place button / link inside form it may cause the _Not found. Authentication passthru_ as well. if you face this issue try removing your button / link outside the form.

**Comment by MarchandMD** [comment]
Also experience this issue. 

I've done the following: 

1. added the `omniauth-rails_csrf_protection` gem (~> 1.0) to the Gemfile
2. changed to a `button_to`
3. updated to `method: :post`
4. added the `session_store.rb` initializer with `Rails.application.config.session_store :cookie_store, key: "_turing_allies_session", domain: "localhost"`
5. added `data: { turbo: false }` to the link options
6. restarted the server

What else am I missing? (oh, and I'm using github to authenticate)

**Comment by MarchandMD** [comment]
While using `omniauth-github` and developing locally, I was experiencing this same issue. 

2 things definitely resolved this for me: 

1. changing `link_to` to `button_to` (even with `method: :post` and `data: {turbo: false}` this was still necessary)
2. updating the callback url in my Oauth app setting on Github specifically. I had `https://localhost:3000/users/auth/github` and I needed to remove the s from https
